### ISSUE TITLE
decoupled_lm support?

### ISSUE THREAD

**Issue by dsvensson** [quakeforge/quakeforge#50]
ericw light recently got support for decoupled lightmaps that allows higher lightmap resolution without messing with texture scale, would be a nice addition,

https://github.com/ericwa/ericw-tools/blob/9760f5b8da9fb359cea9f406d7c273d213744a40/include/common/bspxfile.hh#L123-L139

Artificial example where the back wall has an upscaled texture scale and thus blurs out the lightmap:

![image](https://github.com/quakeforge/quakeforge/assets/302027/4c48f77f-4d94-47d0-9b8d-eba9a30bb2d8)

And with decoupled_lm, the lightmap gets the intended resolution:

![image](https://github.com/quakeforge/quakeforge/assets/302027/7eba2b75-bb86-4b67-9531-c1da2af013b2)

**Comment by taniwha** [comment]
Does this count? ;)
![qf0282](https://github.com/quakeforge/quakeforge/assets/1664867/90ed7e60-ea98-4b36-ba7f-ae1c3f6321a3)

Jokes aside, this seems to be worth looking into when I get light maps working again for vulkan.

**Comment by dsvensson** [comment]
I just added support in ezQuake, so there is one sample implementation. Not added in the QS-lineage of engines yet.

**Comment by dsvensson** [comment]
Wow, the decoupledlm feature was actually implemented for the new Quake 2 remaster now released, more details here:

https://bethesda.net/en/article/6NIyBxapXOurTKtF4aPiF4/enhancing-quake-ii

**Comment by taniwha** [comment]
I plan on getting directional light shadows working (mostly just getting the transforms right), then I'll merge the shadow code into master (despite it being slow for larger maps such as marcher (10fps) or ad_tears(1fps)). I might do some lightmap work after that as I do want vulkan to be playable on big maps.

**Comment by taniwha** [comment]
Related to #73, not not required for it (but might as well look into it while doing so).

**Comment by taniwha** [comment]
@dsvensson Do you  have any sample maps for testing? I've not only got my deferred lighting shadows working, but forward rendering more or less done, so I'm ready to tackle this.

**Comment by dsvensson** [comment]
[testmaps.zip](https://github.com/quakeforge/quakeforge/files/13997476/testmaps.zip)
Here are two test maps with DECOUPLED_LM bspx lump. Lightmaps are only found in RGBLIGHTING bspx lump, original LMs are 0'd due to -novanilla. I've included progsdump's progs.dat as the maps use ents from that (well, actually ktx's qwprogs.so/dll, but a progs.dat of progsdump is probably more compatible). One of the maps also contain the new LIGHTGRID_OCTREE that might be of interest for your realtime lighting. Lightgrid is what it says, provides average lighting for volume segments of the map to correctly light entities.

**Comment by taniwha** [comment]
Thanks. there are some bugs in the vulkan lightmap code (I always get bugs in lightmaps :/), but I've got the forward renderer essentially done (some validation errors with fisheye, but that's more a toy).

**Comment by dsvensson** [comment]
Just tried master, and I'm getting validation errors on startup followed by exit. viewmask vs maxviews, perhaps some missing property probing, I'll update with exact details here when I get back. NVIDIA 1650 (laptop) with 545 drivers, Ubuntu 23.10+x11.

**Comment by taniwha** [comment]
Oh! Nice, I haven't been getting validation errors, but I've had some issues. They may be related.

In regards to decoupled lm, all the ezquake changes are marked with `LIGHTMAPS:`, right?

**Comment by taniwha** [comment]
@dsvensson Ah, I just read through the validation error: not what I was hoping, but it is a known issue (thanks for the reminder): I need to sort out limits in my render-graph config as it currently cannot deal with such. As a workaround, you can comment out all the `shadowN = ...` lines in `libs/video/renderer/vulkan/rp_main_def.plist` where N >= 16 (if I'm reading that message correctly).

**Comment by taniwha** [comment]
Nice map :) (using RT lights. not sure what's up with the frame rate, there aren't that many lights. debug build, but the non-debug build got stuck first try? Seems to be the hexagons in the center (if it's clusters, I have an optimization for that in mind))

![Image](https://github.com/quakeforge/quakeforge/assets/1664867/03a8f58c-c8fd-49a4-877d-0b046abf9029)

**Comment by dsvensson** [comment]
Unfortunately realtime lighting is not available with _minlight as func_detail_* are compiled away :( Maybe I should switch to surface lights to give realtime lighting a chance to draw it correctly. Here's the correct look, https://youtu.be/T5GSHdd2hvk

**Comment by taniwha** [comment]
Very nice. Unfortunately, QF doesn't support the QC extensions needed for the progs (yet).
Hmm, rather than having you switch, maybe just extract the lighting, though adding minlight on a per-surface basis could be interesting. QF parses the bsp's entity string, which can be overridden, including with a plist version.

Anyway, BSPX itself seems to be pretty easy, so getting that going (along with rgb lights) and then decoupled lightmaps shouldn't take too long.

**Comment by dsvensson** [comment]
Yeah, so with _surface=texture property of a light ent, that instead signals making such textures emissive, as used on the second map for all the fluorescent tubes. Think it's also used on the map above for the lava for example, and the teleporters.

**Comment by taniwha** [comment]
akator doesn't work properly if the engine doesn't support detached lightmaps. I wasted a bit of time tracking that one down :/ (but I did get BSPX RGBLIGHTING working).

**Comment by taniwha** [comment]
@dsvensson due to QF's architecture, this is a much bigger project than expected. It really doesn't help that I've been working to optimize the data structures for brush models. At one stage, I discovered msurface_t had gotten to be at least 192 bytes, I've current got it down to 128, so adding lminfo data to it is... not going to happen, so I need to come up with a new way of dealing with lightmap data for the GPU renderers (won't bother with dclm support for software, at least at this stage).

**Comment by taniwha** [comment]
@dsvensson btw, a missed optimization opportunity:

    local[0] = DotProduct (impact, surf->lmvecs[0]) + surf->lmvecs[0][3] - surf->texturemins[0];

Since everything's on the same surf, texturemins[i] can be subtracted from lmvecs[i][3] after setting up the UVs and then the dlight updates can be done with just the add (or better yet, a single vec4 dot product if your impact is `{x,y,z,1}`)
